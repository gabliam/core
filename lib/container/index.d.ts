import { Container, inject } from 'inversify';
declare let container: Container;
declare let provide: any;
declare let fluentProvider: any;
export { container, provide, fluentProvider, inject };
