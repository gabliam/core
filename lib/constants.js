"use strict";
const TYPE = {
    Controller: Symbol("Controller"),
    Service: Symbol("Service"),
    Config: Symbol("Config")
};
exports.TYPE = TYPE;
const METADATA_KEY = {
    controller: "_controller",
    controllerMethod: "_controller-method",
    Bean: "_bean",
};
exports.METADATA_KEY = METADATA_KEY;
const DEFAULT_ROUTING_ROOT_PATH = "/";
exports.DEFAULT_ROUTING_ROOT_PATH = DEFAULT_ROUTING_ROOT_PATH;
