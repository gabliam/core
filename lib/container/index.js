"use strict";
const inversify_1 = require("inversify");
exports.inject = inversify_1.inject;
const inversify_logger_middleware_1 = require("inversify-logger-middleware");
const inversify_binding_decorators_1 = require("inversify-binding-decorators");
let container = new inversify_1.Container();
exports.container = container;
if (process.env.NODE_ENV === 'development') {
    let logger = inversify_logger_middleware_1.makeLoggerMiddleware();
    container.applyMiddleware(logger);
}
let provide = inversify_binding_decorators_1.makeProvideDecorator(container);
exports.provide = provide;
let fluentProvider = inversify_binding_decorators_1.makeFluentProvideDecorator(container);
exports.fluentProvider = fluentProvider;
