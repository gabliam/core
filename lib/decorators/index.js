"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
__export(require("./controller"));
__export(require("./service"));
__export(require("./bean"));
__export(require("./config"));
