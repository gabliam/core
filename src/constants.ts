/** Type in registry */
export const TYPE = {
    Service: 'ServiceType',
    Config: 'ConfigType',
};

export const METADATA_KEY = {
    bean: '_bean',
    value: '_value',
    register: '_register',
    scan: '_scan',
};

export const ORDER_CONFIG = {
    Core: 50,
    Plugin: 100,
    Config: 150,
};

export const APP_CONFIG = Symbol('APP_CONFIG');

export const CORE_CONFIG = Symbol('CORE_CONFIG');
