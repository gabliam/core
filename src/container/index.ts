import { Container } from 'inversify';
import { makeLoggerMiddleware } from 'inversify-logger-middleware';
import { makeValueMiddleware } from './valueMiddleware';

/**
 * Create the inversify container
 */
export function createContainer(): Container {
    let container = new Container();

    let middlewares = [];
    if (process.env.NODE_ENV === 'development') {
        let logger = makeLoggerMiddleware();
        middlewares.push(logger);
    }

    middlewares.push(makeValueMiddleware(container));
    container.applyMiddleware(...middlewares);
    return container;
}

export { Container };
