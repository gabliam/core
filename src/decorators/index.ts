export * from './service';
export * from './bean';
export { Config, PluginConfig } from './config';
export * from './value';
export * from './scan';
export * from './register';
