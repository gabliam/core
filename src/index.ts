import * as interfaces from './interfaces';
export * from './decorators';
export * from './gabliam';
export { APP_CONFIG, CORE_CONFIG } from './constants';

export * from './registry';
export * from './errors';

export {
  injectable, tagged, named, inject, optional, unmanaged, multiInject, targetName, decorate, interfaces as inversifyInterfaces
} from 'inversify';

export { interfaces };
