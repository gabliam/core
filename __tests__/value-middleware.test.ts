// tslint:disable:one-line
// tslint:disable:no-unused-expression
import { createContainer } from '../src/container';
import { APP_CONFIG } from '../src/constants';
import { Config, Value } from '../src/decorators';
import { inject, optional, Container } from '../src/index';
import * as Joi from 'joi';

describe('Unit Test: Value midlleware', () => {
  let container: Container;

  describe('@Value(path: string) Value without others options', () => {
    beforeEach(() => {
      container = createContainer();
      container.bind(APP_CONFIG).toConstantValue({
        application: {
          express: {
            rootPath: '/',
            hostname: 'localhost'
          }
        }
      });
    });
    it('all value in config', () => {
      @Config()
      class Conf {
        @Value('application.express.rootPath') rootPath: string;

        @Value('application.express.hostname') hostname: string;
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      const conf = container.get<Conf>(Conf);
      expect(conf).toMatchSnapshot();
    }); // end all value in config

    it('default value in property', () => {
      @Config()
      class Conf {
        @Value('application.express.rootPath') rootPath: string;

        @Value('application.express.port') port = 3000;

        @Value('application.express.hostname') hostname: string;
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      const conf = container.get<Conf>(Conf);
      expect(conf).toMatchSnapshot();
    }); // end default value in property

    it('with value undefined', () => {
      @Config()
      class Conf {
        @Value('application.express.rootPath') rootPath: string;

        @Value('application.express.port') port = 3000;

        @Value('application.express.lol') lol: string;

        @Value('application.express.hostname') hostname: string;
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      const conf = container.get<Conf>(Conf);
      expect(conf).toMatchSnapshot();
    }); // end whith value undefined

    it('default value in constructor', () => {
      @Config()
      class Conf {
        @Value('application.express.rootPath') rootPath: string;

        @Value('application.express.port') port: number;

        @Value('application.express.hostname') hostname: string;

        constructor() {
          this.port = 8080;
        }
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      const conf = container.get<Conf>(Conf);
      expect(conf).toMatchSnapshot();
    }); // end default value in constructor

    it('default value in constructor and property', () => {
      @Config()
      class Conf {
        @Value('application.express.rootPath') rootPath: string;

        @Value('application.express.port') port = 3000;

        @Value('application.express.hostname') hostname: string;

        constructor() {
          this.port = 8080;
        }
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      const conf = container.get<Conf>(Conf);
      expect(conf).toMatchSnapshot();
    }); // end default value in constructor

    it('default value in other bean', () => {
      container.bind('PORT_CONFIG').toConstantValue(8081);
      @Config()
      class Conf {
        @Value('application.express.rootPath') rootPath: string;

        @Value('application.express.port') port = 3000;

        @Value('application.express.hostname') hostname: string;

        constructor(@inject('PORT_CONFIG') portConfig: number) {
          this.port = portConfig;
        }
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      const conf = container.get<Conf>(Conf);
      expect(conf).toMatchSnapshot();
    }); // end default value in constructor

    it('default value and optional bean', () => {
      @Config()
      class Conf {
        @Value('application.express.rootPath') rootPath: string;

        @Value('application.express.hostname') hostname: string;

        constructor(
          @inject('ROOT_CONFIG')
          @optional()
          rootPath: string
        ) {
          this.rootPath = rootPath;
        }
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      const conf = container.get<Conf>(Conf);
      expect(conf).toMatchSnapshot();
    }); // end default value and optional bean
  }); // end @Value(path: string)

  describe('@Value(path: string, schema: Joi.Schema) Value with Schema', () => {
    beforeEach(() => {
      container = createContainer();
      container.bind(APP_CONFIG).toConstantValue({
        application: {
          express: {
            rootPath: '/',
            hostname: 'localhost'
          }
        }
      });
    });
    it('all value in config and respect Validator', () => {
      @Config()
      class Conf {
        @Value('application.express.rootPath', Joi.string().required())
        rootPath: string;

        @Value('application.express.hostname', Joi.string().required())
        hostname: string;
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      const conf = container.get<Conf>(Conf);
      expect(conf).toMatchSnapshot();
    }); // end all value in config

    it('default value in property', () => {
      @Config()
      class Conf {
        @Value('application.express.rootPath', Joi.string().required())
        rootPath: string;

        @Value('application.express.port', Joi.number().required())
        port = 3000;

        @Value('application.express.hostname', Joi.string().required())
        hostname: string;
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      const conf = container.get<Conf>(Conf);
      expect(conf).toMatchSnapshot();
    }); // end default value in property

    it('bad format', () => {
      @Config()
      class Conf {
        @Value('application.express.rootPath', Joi.number())
        rootPath: string;
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      expect(() => {
        container.get<Conf>(Conf);
      }).toThrow();
    }); // end bad format

    it('error required', () => {
      @Config()
      class Conf {
        @Value('application.express.rootPath', Joi.string())
        rootPath: string;

        @Value('application.express.rootPath', Joi.number().required())
        port: number;
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      expect(() => {
        container.get<Conf>(Conf);
      }).toThrow();
    }); // end error required

    it('default value in constructor', () => {
      @Config()
      class Conf {
        @Value('application.express.rootPath', Joi.string().required())
        rootPath: string;

        @Value('application.express.port', Joi.number().required())
        port: number;

        @Value('application.express.hostname', Joi.string().required())
        hostname: string;

        constructor() {
          this.port = 8080;
        }
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      const conf = container.get<Conf>(Conf);
      expect(conf).toMatchSnapshot();
    }); // end default value in constructor

    it('default in joi', () => {
      @Config()
      class Conf {
        @Value('application.express.rootPath', Joi.string().required())
        rootPath: string;

        @Value('application.express.port', Joi.string().default(8088))
        port: number;

        @Value('application.express.hostname', Joi.string().required())
        hostname: string;
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      const conf = container.get<Conf>(Conf);
      expect(conf).toMatchSnapshot();
    }); // end default value in constructor
  }); // end @Value(path: string, schema: Joi.Schema)

  describe('@Value(options: ValueOptions) Value with Schema', () => {
    beforeEach(() => {
      container = createContainer();
      container.bind(APP_CONFIG).toConstantValue({
        application: {
          express: {
            rootPath: '/',
            hostname: 'localhost'
          }
        }
      });
    });
    it('all value in config and respect Validator', () => {
      @Config()
      class Conf {
        @Value({
          path: 'application.express.rootPath',
          validator: Joi.string().required()
        })
        rootPath: string;

        @Value({
          path: 'application.express.hostname',
          validator: Joi.string().required()
        })
        hostname: string;
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      const conf = container.get<Conf>(Conf);
      expect(conf).toMatchSnapshot();
    }); // end all value in config

    it('default value in property', () => {
      @Config()
      class Conf {
        @Value({
          path: 'application.express.rootPath',
          validator: Joi.string().required()
        })
        rootPath: string;

        @Value({
          path: 'application.express.port',
          validator: Joi.number().required()
        })
        port = 3000;

        @Value({
          path: 'application.express.hostname',
          validator: Joi.string().required()
        })
        hostname: string;
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      const conf = container.get<Conf>(Conf);
      expect(conf).toMatchSnapshot();
    }); // end default value in property

    it('bad format', () => {
      @Config()
      class Conf {
        @Value({
          path: 'application.express.rootPath',
          validator: Joi.number()
        })
        rootPath: string;
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      expect(() => {
        container.get<Conf>(Conf);
      }).toThrow();
    }); // end bad format

    it('bad format custom error', () => {
      @Config()
      class Conf {
        @Value({
          path: 'application.express.rootPath',
          validator: {
            schema: Joi.number(),
            customErrorMsg: 'lol custom'
          }
        })
        rootPath: string;
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      expect(() => {
        container.get<Conf>(Conf);
      }).toThrow('custom');
    }); // end bad format

    it('bad format without throw error', () => {
      @Config()
      class Conf {
        @Value({
          path: 'application.express.rootPath',
          validator: {
            schema: Joi.number(),
            throwError: false
          }
        })
        rootPath: string;
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      const conf = container.get<Conf>(Conf);
      expect(conf).toMatchSnapshot();
    }); // end bad format

    it('error required', () => {
      @Config()
      class Conf {
        @Value({
          path: 'application.express.rootPath',
          validator: Joi.string()
        })
        rootPath: string;

        @Value({
          path: 'application.express.rootPath',
          validator: Joi.number().required()
        })
        port: number;
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      expect(() => {
        container.get<Conf>(Conf);
      }).toThrow();
    }); // end error required

    it('default value in constructor', () => {
      @Config()
      class Conf {
        @Value({
          path: 'application.express.rootPath',
          validator: Joi.string().required()
        })
        rootPath: string;

        @Value({
          path: 'application.express.port',
          validator: Joi.number().required()
        })
        port: number;

        @Value({
          path: 'application.express.hostname',
          validator: Joi.string().required()
        })
        hostname: string;

        constructor() {
          this.port = 8080;
        }
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      const conf = container.get<Conf>(Conf);
      expect(conf).toMatchSnapshot();
    }); // end default value in constructor

    it('default in joi', () => {
      @Config()
      class Conf {
        @Value({
          path: 'application.express.rootPath',
          validator: Joi.string().required()
        })
        rootPath: string;

        @Value({
          path: 'application.express.port',
          validator: Joi.string().default(8088)
        })
        port: number;

        @Value({
          path: 'application.express.hostname',
          validator: Joi.string().required()
        })
        hostname: string;
      }

      container.bind(Conf).to(Conf).inSingletonScope();
      const conf = container.get<Conf>(Conf);
      expect(conf).toMatchSnapshot();
    }); // end default value in constructor
  }); // end @Value(path: string, schema: Joi.Schema)
});
