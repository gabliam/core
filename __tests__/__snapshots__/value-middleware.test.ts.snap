// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@Value(options: ValueOptions) Value with Schema all value in config and respect Validator 1`] = `
Conf {
  "hostname": "localhost",
  "rootPath": "/",
}
`;

exports[`@Value(options: ValueOptions) Value with Schema bad format without throw error 1`] = `
Conf {
  "rootPath": null,
}
`;

exports[`@Value(options: ValueOptions) Value with Schema default in joi 1`] = `
Conf {
  "hostname": "localhost",
  "port": 8088,
  "rootPath": "/",
}
`;

exports[`@Value(options: ValueOptions) Value with Schema default value in constructor 1`] = `
Conf {
  "hostname": "localhost",
  "port": 8080,
  "rootPath": "/",
}
`;

exports[`@Value(options: ValueOptions) Value with Schema default value in property 1`] = `
Conf {
  "hostname": "localhost",
  "port": 3000,
  "rootPath": "/",
}
`;

exports[`@Value(path: string) Value without others options all value in config 1`] = `
Conf {
  "hostname": "localhost",
  "rootPath": "/",
}
`;

exports[`@Value(path: string) Value without others options default value and optional bean 1`] = `
Conf {
  "hostname": "localhost",
  "rootPath": "/",
}
`;

exports[`@Value(path: string) Value without others options default value in constructor 1`] = `
Conf {
  "hostname": "localhost",
  "port": 8080,
  "rootPath": "/",
}
`;

exports[`@Value(path: string) Value without others options default value in constructor and property 1`] = `
Conf {
  "hostname": "localhost",
  "port": 8080,
  "rootPath": "/",
}
`;

exports[`@Value(path: string) Value without others options default value in other bean 1`] = `
Conf {
  "hostname": "localhost",
  "port": 8081,
  "rootPath": "/",
}
`;

exports[`@Value(path: string) Value without others options default value in property 1`] = `
Conf {
  "hostname": "localhost",
  "port": 3000,
  "rootPath": "/",
}
`;

exports[`@Value(path: string) Value without others options with value undefined 1`] = `
Conf {
  "hostname": "localhost",
  "lol": undefined,
  "port": 3000,
  "rootPath": "/",
}
`;

exports[`@Value(path: string, schema: Joi.Schema) Value with Schema all value in config and respect Validator 1`] = `
Conf {
  "hostname": "localhost",
  "rootPath": "/",
}
`;

exports[`@Value(path: string, schema: Joi.Schema) Value with Schema default in joi 1`] = `
Conf {
  "hostname": "localhost",
  "port": 8088,
  "rootPath": "/",
}
`;

exports[`@Value(path: string, schema: Joi.Schema) Value with Schema default value in constructor 1`] = `
Conf {
  "hostname": "localhost",
  "port": 8080,
  "rootPath": "/",
}
`;

exports[`@Value(path: string, schema: Joi.Schema) Value with Schema default value in property 1`] = `
Conf {
  "hostname": "localhost",
  "port": 3000,
  "rootPath": "/",
}
`;
